### PR TITLE
SmsServiceAgent + NuggetPackage Origam-sms

### PR DESCRIPTION
**Pull request by MikulaJan**
Origam.Workflow.ServiceAdapters 
Added SmsService adapter with dependency on Origam-sms Nugetpackage.

### CODE REVIEW

**Review by washibana — CHANGES_REQUESTED** [review]
Files in `.vs` folder shouldn't be part of the commit.

I'm not convinced that `DebugSmsService` is the correct name. I guess it stores messages on the filesystem. If I'm right, I'd name it `FileSystemSmsService`.

SmsServiceAdapter - Method `Run` doesn't perform any usual input checks, whether the mandatory input parameters are filled.

SmsService should be provided externally, most likely via setter (we don't want to change the workflow agent invocation to much, so this is good way to close to DI).

`ServiceAgentFactory` - method `GetAgent` should have an entry for SmsService.

Model should be updated to support SmsService.

**Comment by tvavrda** [comment]
ServiceAgentFactory - method GetAgent should have an entry for SmsService.

I do not agree here. Services should be generally registered via an assembly name in model. It is not necessary to change a code.

    SmsServiceAdapter - Method Run doesn't perform any usual input checks, whether the mandatory input parameters are filled.

Checks are provided by calling base `GetParameter<T>` calls, which check for the type and require the parameter to be not null (for non-mandatory parameters one would call `TryGetParameter<T>`

    Model should be updated to support SmsService.

Correct, the question is if we should put it straight into Root package or create a separate one?

File: backend/Origam.Workflow/Service Agents/SmsServiceAgent.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        }
+        public override void Run()
+        {
+            switch (MethodName)
+            {
+                case "SendSms":
+                    _result = CreateSmsService().SendSms(
+                        GetParameter<string>("from"),
+                        GetParameter<string>("to"),
+                        GetParameter<string>("to"));
```
Comment (tvavrda) [inline backend/Origam.Workflow/Service Agents/SmsServiceAgent.cs:48]: This should probably be `body`, not `to`?

**Review by tvavrda — CHANGES_REQUESTED** [review]

**Comment by washibana** [comment]
> ```
> ServiceAgentFactory - method GetAgent should have an entry for SmsService.
> ```
> 
> I do not agree here. Services should be generally registered via an assembly name in model. It is not necessary to change the code.
> 
OK, I expected it to be made the same way as MailService, which is handled by ServiceAgentFactory. In that case I don't think we have means to inject SmsService into the adapter.

> 
> ```
> Model should be updated to support SmsService.
> ```
> 
> Correct, the question is if we should put it straight into Root package or create a separate one?

Is there a reason to have it in root except less work for those who want to use SmsService?

**Comment by tvavrda** [comment]
In general service agents should be "plugins" so our code should not be responsible for handling anything inside the adapter. So it creates an instance inside by using reflection. That's the developer's choice, not ORIGAM's. 

As for Root model - there is no reason, except if we use it inside Root, which we don't.

**Comment by tvavrda** [comment]
Mail service was a different use case - we had to support 2 platforms. It will go away as soon as we move fully to .net core. It was a temporary solution, not really systematic, in my opinion, but it worked.

**Comment by washibana** [comment]
In such case, plugins should have their own config files.

**Comment by washibana** [comment]
> In general service agents should be "plugins" so our code should not be responsible for handling anything inside the adapter. So it creates an instance inside by using reflection. That's the developer's choice, not ORIGAM's.

Shouldn't then the plugin use its own config file instead of OrigamSettings? It would also mean to drop its dependency on OrigamSettings.

File: backend/Origam.Workflow/Service Agents/SmsServiceAgent.cs (hunk 1 above)
Comment (MikulaJan) [reply backend/Origam.Workflow/Service Agents/SmsServiceAgent.cs:48]: Yes, I was talking about it at a recent call, weird I committed older. Will Fix

**Comment by tvavrda** [comment]
Shouldn't then the plugin use its own config file instead of OrigamSettings? It would also mean to drop its dependency on OrigamSettings.

Agree. Typically they should not, at least those not distributed with ORIGAM directly. This is such a case though and so in order to keep the style (like DataService config) I decided to put it there. On the other hand I feel the time to say good bye to OrigamSettings is close...

**Review by washibana — APPROVED** [review]

**Pull request merged**